source/slides/04_troubleshooting_bash.rst

```diff
@@ -0,0 +1,237 @@
+.. 04_troubleshooting_bash
+
+Troubleshooting and Bash
+========================
+
+Troubleshooting
+---------------
+
+* This is a lot like debugging a problem in code
+* 
+
+General Steps
+-------------
+
+#. Understand the problem
+#. Reproduce the problem
+#. Isolate the change that created the problem
+#. Test the fix
+#. Automate the fix to your entire infrastructure
+
+Example Problem
+---------------
+
+* Nagios alert goes off that SSH isn't working
+* Confirm SSH isn't working
+* Log in over serial via the BMC
+
+Symptoms
+--------
+
+* Almost all executables return ``Input/Output Error``
+* Only working (useful) utils: ``cat``, ``ssh``, bash builtins
+* Can't read most files (only ones that are already in memory)
+* Ideas..?
+
+(Temporary) Fixing
+------------------
+
+* A reboot (temporarily) fixed the issue
+* No logs (couldn't write to disk)
+* No useful logs went to remote servers
+* No logs in the Server Event Log (BMC log)
+* Raid card reports 1 disk that will fail soon
+
+Possible Causes
+---------------
+
+* Likely a RAID card lockup
+* Bad controller?
+* Bad disk?
+* Buggy controller?
+* Bad PCIE bridge? (nope)
+
+Real Fix
+--------
+
+* Contact vendor and start RMA process for controller
+* See if problem reproduces
+
+Example Problem 2
+-----------------
+
+* Google Analytics reports a growing number of 404 links
+* Links are growing exponentially (small number, 1k, 9k, 50k, ...)  per day
+* None of the reported links can be found on the site.
+
+Symptoms
+--------
+
+* Link growth pattern
+* Reported links sometimes have patterns, ``/path/page/2, /path/page/3, ...``
+
+Debugging
+---------
+
+* Change browser string to be the same as Google Bot
+* Nothing changes
+
+* Use ``curl`` to inspect page at a lower level, notice temporary redirects
+* Browsers follow redirect without reading page, bots read page.
+
+Cause
+-----
+
+* Every temporary redirect page was still generating page content
+
+  * Including links that were invalid (404!)
+
+Fix
+---
+
+* Changed webapp to use permanent redirect, not temporary
+* Wait for GA 404 reports to go back down.
+
+Bash
+====
+
+Basics
+------
+
+* Interpreted Language
+* Built to easily interact with a system, run other programs
+* Pipes!
+
+Useful Symbols
+--------------
+
+* ``|`` Pipe -- redirects stdout of left to stdin of right
```

Comment:
Its rather hard to read this slide. I wonder if it might work better using an entire bash code block with comments as the content. Thoughts?